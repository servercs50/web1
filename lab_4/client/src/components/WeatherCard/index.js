import React from 'react';

import {Grid, Paper, Fab, Hidden} from "@material-ui/core";
import Typography from '@material-ui/core/Typography/index';
import CloseIcon from '@material-ui/icons/Close';

import LoadingSpinner from '../LoadingSpinner';

import './style.sass';


export default function WeatherCard(props) {
  const data = props.cityWeatherData;;

  if (data && data.isLoading) {
    return (<LoadingSpinner/>);
  }

  if (data && data.cod && data.cod != 200) {
    if ('onFetchError' in props) props.onFetchError()
    return (
      <div>
        <p>К сожалению, не получилось получить данные о погоде...</p>
        <p>{data.message}</p>
      </div>
    )
  }

  if (data && data.cod == 200) {
    const K = 273.15; // для перевода Кельвин в Цельсия
    const temperatureCelsius = Math.round(data.main.temp - K);

    return (
      <Grid container direction={props.isFavorite ? "column" : "row"} spacing={4}
            className={`WeatherCard ${props.isFavorite ? "isFavorite" : ""}`}>

        <Grid container item xs={12} md={props.isFavorite ? 12 : 6}
              direction={props.isFavorite ? "row" : "column"} alignItems={"center"} className={"weather-main"}>
          <Grid container item sm={props.isFavorite ? 4 : false} justify={"space-between"}>
            <Typography variant="h5" component="h2">
              <b>{data.name}</b>
            </Typography>
            <Hidden smUp>
              {props.isFavorite &&
              <Fab color="primary" size={"small"}
                   onClick={props.onRemoveCityClick.bind(this, props.cityWeatherData.name)}>
                <CloseIcon/>
              </Fab>
              }
            </Hidden>
          </Grid>
          <Grid item container sm={props.isFavorite ? 7 : false} alignItems={"center"}
                justify={"space-evenly"} className={"weather-temperature"}>
            <img src={`http://openweathermap.org/img/wn/${data.weather[0].icon}@2x.png`} alt=""/>
            <Typography variant={props.isFavorite ? "h3" : "h1"} component="h2">
              {temperatureCelsius}°C
            </Typography>
          </Grid>
          <Hidden only="xs">
            <Grid container item sm={1} justify={"flex-end"}>
              {props.isFavorite &&
              <Fab color="primary" size={"small"}
                   onClick={props.onRemoveCityClick.bind(this, props.cityWeatherData.cityName)}>
                <CloseIcon/>
              </Fab>
              }
            </Grid>
          </Hidden>
        </Grid>

        <Grid item xs={12} md={props.isFavorite ? 12 : 6} className={"weather-items"}>
          <Paper elevation={1} className={"weather-item"}>
            <span><b>Ветер</b></span>
            <span>{data.wind.speed} м/с</span>
          </Paper>

          <Paper elevation={1} className={"weather-item"}>
            <span><b>Облачность</b></span>
            <span>{data.weather[0].description}</span>
          </Paper>

          <Paper elevation={1} className={"weather-item"}>
            <span><b>Давление</b></span>
            <span>{data.main.pressure} hpa</span>
          </Paper>

          <Paper elevation={1} className={"weather-item"}>
            <span><b>Влажность</b></span>
            <span>{data.main.humidity} %</span>
          </Paper>

          <Paper elevation={1} className={"weather-item"}>
            <span><b>Координаты</b></span>
            <span>[{data.coord.lon}, {data.coord.lat}]</span>
          </Paper>
        </Grid>
      </Grid>
    )
  }

  return null;
}
