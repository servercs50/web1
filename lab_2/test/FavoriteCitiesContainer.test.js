import React from 'react';
import {shallow} from 'enzyme';
import {shallowToJson} from 'enzyme-to-json';

import {FavoriteCities} from '../src/components/FavoriteCitiesContainer';

const cityWeatherData = {
  "coord": {"lon": 37.62, "lat": 55.75},
  "weather": [{"id": 803, "main": "Clouds", "description": "broken clouds", "icon": "04n"}],
  "base": "stations",
  "main": {"temp": 283.57, "pressure": 995, "humidity": 87, "temp_min": 282.59, "temp_max": 285.15},
  "visibility": 10000,
  "wind": {"speed": 9, "deg": 230},
  "clouds": {"all": 75},
  "dt": 1572987525,
  "sys": {"type": 1, "id": 9027, "country": "RU", "sunrise": 1573015504, "sunset": 1573047671},
  "timezone": 10800,
  "id": 524901,
  "name": "Moscow",
  "cod": 200
}


describe('Favorite Cities Container', () => {

  it('looks for cities in localstorage', () => {

    const fetchCityProp = jest.fn()
    fetchCityProp.mockReturnValueOnce(cityWeatherData)
    const removeCityProp = jest.fn()

    const tree = shallow(
      <FavoriteCities
        cities={[{name: 'Moscow'}, {name: 'London'}]}
        fetchCity={fetchCityProp}
        removeCity={removeCityProp}
      />
    );

    expect(shallowToJson(tree)).toMatchSnapshot();
    expect(fetchCityProp).toHaveBeenCalledTimes(2)
    expect(fetchCityProp).toHaveBeenCalledWith('Moscow', true)
    expect(fetchCityProp).toHaveBeenCalledWith('London', true)
  });

})
