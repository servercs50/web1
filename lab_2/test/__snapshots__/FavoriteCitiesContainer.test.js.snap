// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Favorite Cities Container looks for cities in localstorage 1`] = `
<WithStyles(ForwardRef(Grid))>
  <WithStyles(ForwardRef(Grid))
    alignItems="center"
    component="section"
    container={true}
  >
    <WithStyles(ForwardRef(Grid))
      item={true}
      sm={5}
      xs={12}
    >
      <WithStyles(ForwardRef(Typography))
        component="h2"
        variant="h4"
      >
        Избранное
      </WithStyles(ForwardRef(Typography))>
    </WithStyles(ForwardRef(Grid))>
    <WithStyles(ForwardRef(Grid))
      className="addCityInput"
      container={true}
      item={true}
      sm={7}
      xs={12}
    >
      <form
        onSubmit={[Function]}
      >
        <WithStyles(ForwardRef(TextField))
          label="Добавить новый город"
          margin="none"
          name="city"
          onChange={[Function]}
          placeholder="Город"
          value=""
        />
        <WithStyles(ForwardRef(Fab))
          color="primary"
          size="small"
          style={
            Object {
              "marginLeft": "1rem",
              "marginTop": "0.5rem",
            }
          }
          type="submit"
        >
          <AddIcon />
        </WithStyles(ForwardRef(Fab))>
      </form>
    </WithStyles(ForwardRef(Grid))>
  </WithStyles(ForwardRef(Grid))>
  <WithStyles(ForwardRef(Grid))
    component="section"
    container={true}
    spacing={4}
  >
    <WithStyles(ForwardRef(Grid))
      item={true}
      key="0"
      md={6}
      xs={12}
    >
      <WeatherCard
        cityWeatherData={
          Object {
            "name": "Moscow",
          }
        }
        isFavorite={true}
        onFetchError={[Function]}
        onRemoveCityClick={[Function]}
      />
    </WithStyles(ForwardRef(Grid))>
    <WithStyles(ForwardRef(Grid))
      item={true}
      key="1"
      md={6}
      xs={12}
    >
      <WeatherCard
        cityWeatherData={
          Object {
            "name": "London",
          }
        }
        isFavorite={true}
        onFetchError={[Function]}
        onRemoveCityClick={[Function]}
      />
    </WithStyles(ForwardRef(Grid))>
  </WithStyles(ForwardRef(Grid))>
</WithStyles(ForwardRef(Grid))>
`;
